propose fix for issue 61

zh3nl opened the pull request:
Fix: For fget, fset, and fdel, I suggest binding the relevant methods to the class. Also use a descriptor protocol for class members. 

Tests: I have created the following tests to test my implementation against read-only class property, deletable class property, class properties with default values, and class property with custom getter/setter methods. 

Please provide feedback as I am a beginner to this!

VigneshVSV commented on hololinked/param/parameterized.py:
```diff
@@ -1795,7 +1805,12 @@ def __setattr__(mcs, attribute_name : str, value : typing.Any) -> None:
+                    # For class member properties, use the descriptor protocol
+                    parameter.__set__(None, value)
+                else:
+                    mcs.__dict__[attribute_name].__set__(mcs, value)
```
This code is from me and could be refactored as `parameter.__set__(mcs, value)` for better readability

zh3nl replied:
Refactor completed.

VigneshVSV commented on hololinked/param/parameterized.py:
```diff
@@ -367,6 +367,9 @@ def __get__(self, obj : typing.Union['Parameterized', typing.Any],
         if self.class_member:
+            if self.fget is not None:
+                 # For class properties, bind the getter to the class
+                 return self.fget.__get__(None, objtype)(objtype)
```
This works because python `FunctionType` itself is a descriptor?

zh3nl replied:
It essentially binds the getter method to the metaclass if no getter function is available. This allows properties to be defined at the class level rather than the instance level. Python FunctionType itself I believe is a descriptor (but correct me if I am wrong).

VigneshVSV replied:
Yes statement 1 is true because FunctionType is a descriptor. I asked because I didn't think of solving it like this, but this should actually work. I will still have a look with the debugger.

VigneshVSV commented on hololinked/param/parameterized.py:
```diff
@@ -418,7 +421,11 @@ def __set__(self, obj : typing.Union['Parameterized', typing.Any], value : typin
-            self.fset(obj, value) 
+            if self.class_member:
+                # For class properties, bind the setter to the class
+                self.fset.__get__(None, obj)(obj, value)
```
same question as above.

zh3nl replied:
This binds the setter method to the class (similar to above).

VigneshVSV commented:
I will also do a step-by-step debug and get back in about a day! Thanks a lot.

VigneshVSV commented on tests/test_property.py:
```diff
@@ -275,6 +275,128 @@ def test_6_pydantic_model_property(self):
+        return getattr(cls, '_managed_value', 0)
+    
+    @managed_class_prop.setter
+    def managed_class_prop(cls, value):
```
This comment was a mistake, we need to revert it and instead change the name of getter and setter. This problem can be tackled in a new issue as the number of changes are more.

VigneshVSV replied:
```python
    @managed_class_prop.getter
    def get_managed_class_prop(cls):
        return getattr(cls, '_managed_value', 0)

    @managed_class_prop.setter
    def set_managed_class_prop(cls, value):
```

This is the syntax. Similar for deleter.

VigneshVSV commented on tests/test_property.py:
```diff
@@ -275,6 +275,128 @@ def test_6_pydantic_model_property(self):
+    
+    @managed_class_prop.setter
+    def managed_class_prop(cls, value):
+        if value < 0:
```
This is alright, but technically, Number property takes a `bounds` value, like `Number(bounds=(0,None), .... )` so that this validation can be skipped.

VigneshVSV commented on tests/test_property.py:
```diff
@@ -275,6 +275,128 @@ def test_6_pydantic_model_property(self):
+        # Test instance-level validation
+        instance = TestClassPropertyThing(instance_name='test3')
+        with self.assertRaises(ValueError):
+            instance.managed_class_prop = -20
```
Could you also assert the instance based access like line 334 & 335 and also repeat 338 & 339

VigneshVSV commented on tests/test_property.py:
```diff
@@ -275,6 +275,128 @@ def test_6_pydantic_model_property(self):
+        self.assertEqual(TestClassPropertyThing.deletable_class_prop, 200)
+        del instance.deletable_class_prop
+        self.assertEqual(TestClassPropertyThing.deletable_class_prop, 100)  # Should return to default
 
```
Also `self.assertEqual(instance.deletable_class_prop, 100)` would be great!

VigneshVSV commented on tests/test_property.py:
```diff
@@ -275,6 +275,128 @@ def test_6_pydantic_model_property(self):
+    
+    # Deletable class property
+    deletable_class_prop = Number(class_member=True, default=100)
+    
```
Temporarily use different names for getter/setter/deleters - https://github.com/hololinked-dev/hololinked/issues/63

VigneshVSV commented:
Also the version in [here](https://github.com/hololinked-dev/hololinked/blob/main/hololinked/__init__.py) and [here](https://github.com/hololinked-dev/hololinked/blob/main/setup.py) can be changed to 0.2.10 so that once this is merged, I can release it to pip and anaconda.

zh3nl commented:
@VigneshVSV Just pushed the following changes:

1. Reverted changes to decorators to the original commit
2. Asserted instance-based access as you suggested. 
3. Added self.assertEqual(instance.deletable_class_prop, 100)

VigneshVSV commented on hololinked/param/parameterized.py:
```diff
@@ -453,6 +460,9 @@ def __set__(self, obj : typing.Union['Parameterized', typing.Any], value : typin
 
     def __delete__(self, obj : typing.Union['Parameterized', typing.Any]) -> None:
```
For this to be invoked, I just found that `ParameterizedMetaclass` requires to have a `__delattr__` to be defined. Otherwise the descriptor itself will be deleted when you call `del` operator. 

Try doing `python -m unittest` or `python test_property.py` to see the behaviour.

VigneshVSV commented on hololinked/param/parameterized.py:
```diff
@@ -418,7 +421,11 @@ def __set__(self, obj : typing.Union['Parameterized', typing.Any], value : typin
+            else:
+                self.fset(obj, value) 
         else: 
             obj.__dict__[self._internal_name] = value
```
I also found out that this has to be 

```
else: 
        if self.class_member:
                setattr(obj, self._internal_name, value)
        else:
            obj.__dict__[self._internal_name] = value
```
or something like that. This will reinvoke the `__setattr__` of `ParameterizedMetaclass` however it will take a different path and call type.__setattr__ instead of resetting the parameter. This is because the `self._internal_name` is different than the parameter's name (or property's name)

you can try it with `python test_property.py` in the tests folder.

I could be wrong. In that case, you can propose what is right.

VigneshVSV commented:
To run the full unit tests, you can do

```bash
cd tests
python -m unittest
```

It needs to pass before I can merge the branch. Unfortunately , for non-admin, its not automatically triggered.

zh3nl commented:
I have attempted to change my code to pass the unit tests but I observe that the set and del methods are not working because they are trying to change the class which will result in "TypeError: 'mappingproxy' object does not support item assignment". This may or may not be in the scope of the issue and I do not want to change too much of the code.

VigneshVSV commented:
> I have attempted to change my code to pass the unit tests but I observe that the set and del methods are not working because they are trying to change the class which will result in "TypeError: 'mappingproxy' object does not support item assignment". This may or may not be in the scope of the issue and I do not want to change too much of the code.

Not a problem! 

This snippet should fix that 

```
else: 
    if self.class_member:
        setattr(obj, self._internal_name, value)
    else:
        obj.__dict__[self._internal_name] = value
```

However, I can take it from here.

VigneshVSV commented:
Also, thanks a lot for the contribution!

zh3nl commented:
I have implemented the changes you suggested. If you want to take it from here, please do so and merge my additions as needed.

VigneshVSV commented:
> I have implemented the changes you suggested. If you want to take it from here, please do so and merge my additions as needed.

Thanks again! Its looking pretty good.

zh3nl commented:
I want to note that in #62 in the chunk between 413 and 420 on parametrized.py I did use getattr rather than obj.__dict__.get. Not sure if will affect anything.

VigneshVSV commented:
> I want to note that in #62 in the chunk between 413 and 420 on parametrized.py I did use getattr rather than obj.**dict**.get. Not sure if will affect anything.

Logically its fine! Its probably just a mere few nanoseconds faster to just use the `__dict__`. The getattr will ultimately look into the `__dict__` if it does not find it anywhere else. However, to account for `mappingproxy` what you did may be more accurate.

zh3nl commented:
Sounds good! I thought there was a slight difference between the two, but wanted to make sure that I wouldn't break anything :) Have a great rest of your day, and thanks for letting me contribute to the project!

VigneshVSV commented:
Hi, I published it to pip. Thanks a lot! 

I had to make one change which was, I reset `self.fget.__get__(None, obj)(obj)` to `self.fget(obj)` & similar for setters and deleters because when you try to bind it using `None` as the first argument, it results in the same object. 
In other words `self.fget.__get__(None, obj)` is exactly equal to `self.fget(obj)`.

You can have a look here: https://pypi.org/project/hololinked/

zh3nl commented:
Okay got it. So this change has the same functionality but is more readable?

VigneshVSV commented:
> Okay got it. So this change has the same functionality but is more readable?

It just avoids one extra function call. It does not make it more or less readable as if you try `self.fget.__get__(None, cls) == self.fget`, you will get `True`.
